### PR TITLE
spike/feat(FX-4798): Try to reduce delay between save/unsave and toast message

### PR DESCRIPTION
**Pull request by dimatretyak**
This PR resolves [FX-4798] 

#### iOS
| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/artsy/eigen/assets/3513494/a42954cd-56b2-4ae0-a79f-f759142277bb" /> | <video src="https://github.com/artsy/eigen/assets/3513494/84099a9a-4ff2-4bd3-aaf1-e8193c3fd327" /> | 

#### Android
| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/artsy/eigen/assets/3513494/a00b85a3-0a35-4f83-8429-057c1ce4670b" /> | <video src="https://github.com/artsy/eigen/assets/3513494/25398007-4879-4a76-bc7c-bd8ee108a4d6" /> | 

### PR Checklist

- [x] I have tested my changes on **iOS** and **Android**.
- [ ] I hid my changes behind a **[feature flag]**, or they don't need one.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [ ] I added an **[app state migration]**, or my changes do not require one.
- [ ] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [x] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

-

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

- Try to reduce delay between save/unsave and toast message - dimatretyak

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md

[FX-4798]: https://artsyproduct.atlassian.net/browse/FX-4798?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/app/utils/mutations/useSaveArtwork.ts [hunk 1]
```diff
@@ -34,14 +35,16 @@ export const useSaveArtwork = ({
       optimisticUpdater: (store) => {
         const artwork = store.get(id)
         artwork?.setValue(nextSavedState, "isSaved")
+
+        optimisticUpdater?.(nextSavedState, store)
```
[the remark this answers is not in the document]
Comment (dimatretyak) [reply src/app/utils/mutations/useSaveArtwork.ts:39]: > **Warning**
> Toast messages may be duplicated sometimes 🥲


https://github.com/artsy/eigen/assets/3513494/9d358dce-3ffe-4a01-91fe-f605414f15d2

[the remark this answers is not in the document]
Comment (dimatretyak) [reply src/app/utils/mutations/useSaveArtwork.ts:39]: cc: @dblandin

**Comment by dimatretyak** [comment]
I see only one single "unpleasant" problem that we can face. Consider an example:
* User click the 💙 button
* The "Artwork Saved" toast message is displayed immediately, **but** the save to the default ("Saved Artworks") artwork list mutation is in progress in the background
* User clicks "Add to a List" CTA in toast
* We display "Select lists for this artwork" bottom sheet
* Getting all the user's artwork lists
* After we have received all the user's artwork lists, we display them. 
* The save artwork to the default ("Saved Artworks") artwork list mutation is **still in progress** in the background. For this reason, the "Saved Artworks" list is displayed as **not** selected
*

File: src/app/utils/mutations/useSaveArtwork.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dimatretyak) [reply src/app/utils/mutations/useSaveArtwork.ts:39]: I'm trying to understand the reason why toast messages are duplicated and fix it 🕺

[the remark this answers is not in the document]
Comment (dimatretyak) [reply src/app/utils/mutations/useSaveArtwork.ts:51]: Fixed here: https://github.com/artsy/eigen/pull/8815/commits/e372314d53ed2b7beaca02daab80df8a8c7b04bc

[the remark this answers is not in the document]
Comment (dimatretyak) [reply src/app/utils/mutations/useSaveArtwork.ts:51]: ### Description
`optimisticupdater` can be called twice for the same mutation (similar [problem](https://github.com/facebook/relay/issues/3060)). For this reason, toast messages that were already displayed could be displayed again

This hack fixes this problem: https://github.com/artsy/eigen/commit/e372314d53ed2b7beaca02daab80df8a8c7b04bc

### Demo
| Before | After |
| ------------- | ------------- |
| <video src="https://github.com/artsy/eigen/assets/3513494/c0998d8d-787f-4810-acdd-8296f852ba6d" /> | <video src="https://github.com/artsy/eigen/assets/3513494/e5ad1d69-4f9c-4f88-b6fa-1c494790d505" /> |

**Comment by dimatretyak** [comment]
Preparing new beta-builds with these changes 🚧

File: src/app/utils/mutations/useSaveArtwork.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (dimatretyak) [reply src/app/utils/mutations/useSaveArtwork.ts:65]: there are still some problems with the 💙 button in production mode on IOS (there is no such problem in staging/development mode) 🥲🥲🥲

https://github.com/artsy/eigen/assets/3513494/b4abe490-ba8c-4618-95da-4f060d3697bf

**Review by anandaroop — COMMENTED** [review]
Seems fine. I have some minor concerns about the added complexity vs the payoff, but curious what others think.

Also, wondering if there's any risk with optimistic update + CTA in the toast. Is it possible to tap "Add to a List" too soon (i.e. before the commit has actually succeeded in the background?), and if so what would the expected behavior be?

File: src/app/Components/ArtworkLists/useSaveArtworkToArtworkLists.ts [hunk 2]
```diff
@@ -17,6 +18,7 @@ export const useSaveArtworkToArtworkLists = (options: Options) => {
   const { onSave, dispatch } = useArtworkListsContext()
   const { artworkListID, removedArtworkIDs } = useArtworkListContext()
   const artwork = useFragment(ArtworkFragment, artworkFragmentRef)
+  const prevSavedState = useRef<boolean | null>(null)
```
Comment (dimatretyak) [inline src/app/Components/ArtworkLists/useSaveArtworkToArtworkLists.ts:21]: do you think it is better to use this approach or [another](https://github.com/artsy/eigen/commit/a131b3c29f7ae3e9b5e1b8b92d89860094214d9d)?

**Comment by dimatretyak** [comment]
> we don't include this for the upcoming release but keep open for further exploration

yes, we came to the same decision with @anandaroop and @nickskalkin

**Comment by dimatretyak** [comment]
> I did experience some duplicate toasts

in which version did you faced this? Could you please [another approach](https://github.com/artsy/eigen/pull/8815/files#r1222201392) in this build 8.14.0 (1674645331)?

**Comment by dblandin** [comment]
Looks more stable in the recent build! Perhaps let's merge this, cut a new beta from main, and have the team run through some QA for the feature, now that many mprovements are merged, this afternoon?

**Review by dblandin — APPROVED** [review]

**Pull request merged**